### PR TITLE
[batch2] add benchmark of mysql db

### PR DESCRIPTION
**Pull request by jigold**

### CODE REVIEW

**Comment by cseed** [comment]
This is super useful, thanks @jigold!  A few high level comments:

 - I'd love to have this checked in, but I don't think it should be part of the regular tests, esp. when they run against the production database and this is designed to find/stress the limits of the database.
 - Also, this seems most useful for benchmarking different database configuration and settings, and we don't want to vary the production database (and in some cases, we can't, like decreasing the database size).
 - Therefore, I think we just have a module you can run that takes a database connection settings and n_jobs, batch_size, batch_parallelism and number of replicates, and runs the benchmark, not integrated with the build system.  And .sql files to create/clean up tables.  When we want to run it, we can just clone the repo and run it directly.  Then we can think about wrapping it in a larger test to spinning up database instances with various node and disk sizes and MySQL settings and see how they perform.
 - You explore number of jobs and batch size, but I think you should also measure amount of batch insert parallelism.  You can use bounded_gather I sent you.  Then basically these two tests correspond to parallelism=1 and parallelism=infinity.
 - I think you can get rid of the pymysql version.  No reason the async version should perform differently, no?

File: build.yaml [hunk 1]
```diff
@@ -58,6 +58,15 @@ steps:
...
     - base_image
+ - kind: buildImage
+   name: db_benchmark_image
+   dockerFile: batch2/benchmark/Dockerfile
+   contextPath: .
+   publishAs: db-benchmark-image
+   dependsOn:
+     - service_base_image
+   scopes:
+    - dev
```
Comment (jigold) [inline build.yaml]: This is set to dev only.

File: build.yaml [hunk 2]
```diff
@@ -663,6 +731,23 @@ steps:
...
+      name: db-benchmark
+      for: completed
+      timeout: 300
+   dependsOn:
+    - default_ns
+    - db_benchmark_image
+    - batch2_database
+    - create_batch2_tables
+   scopes:
+    - dev
```
Comment (jigold) [inline build.yaml]: Set to dev only.

Comment (jigold) [reply build.yaml]: Sorry. Read further down that this should be separate from the build system.

**Comment by jigold** [comment]
I think we should have the following design that runs the benchmarks in k8s because then we are using google's internal network to transmit data (compared to running on my local computer via a cloud proxy):

- Have a `db-benchmark` namespace in k8s specifically for this.

1. create_db.py
a. This will take the parameters needed for `gcloud sql instances create` including database flags, disk space, cores, etc. and create an instance
b. Get the IP address of the instance (hopefully the REST API works for this)
c. Create a database
d. Create user and password for the database
e. Create config file
f. Create secret in the db-benchmark namespace from the config file
 
2. run.py
a. Build the docker image with the benchmark.py code and installs aiomysql, etc.
b. Create pod which mounts the correct secret with the sql config for the instance to use. Environment variables specify the n_replicates, etc. Print out the pod name.
c. Wait for the pod to complete (you have code in CI that does this)
d. Download logs
e. Delete the pod

3. cleanup.py
a. Delete mysql instance
b. Delete kubernetes secret in db-benchmark namespace

Thoughts? 

I tried to think about how to use the current build system and what I would do is add a new CreateSQLInstance step, CreateDatabase takes the instance name and IP address as a parameter, and have CI take a path to the build.yaml file to build from. But this wasn't straightforward with how to do this, so I thought the above was simpler to reason about.

**Comment by jigold** [comment]
This should work. I didn't test the `cleanup-db` because I wanted to leave the db I have running. Right now, you have to run this from inside the db-benchmark directory. I'm not sure how to easily make it directory agnostic and didn't want to put the time into that now.

```
cd db-benchmark
pip3 install -U ./

db-benchmark create-db --tier db-n1-standard-1 test-ad914f  # this will assign a random name for the db if not specified 

db-benchmark run test-ad914f --parallelism 5 --batch-sizes 1,10,100,1000,10000,100000 --chunk-size 1000 --n-replicates 10

db-benchmark cleanup-db  test-ad914f
```

By default, the k8s logs go to `benchmark.log`. I couldn't get them to print out nicely to the console.

I'll try testing some of the database flags tomorrow and see if any of the suggestions on stack overflow help.

**Pull request closed without merging**